pulsar-broker/src/test/java/org/apache/pulsar/broker/service/OneWayReplicatorTest.java

```diff
@@ -153,4 +179,203 @@ public void testTopicCloseWhenInternalProducerCloseErrorOnce() throws Exception
+        //   0.1s, 0.2, 0.4, 0.8, 1.6s, 3.2s, 6.4s...
+        //   If the retry counter is larger than 6, the next creation will be slow enough to close Replicator.
+        final AtomicInteger createProducerCounter = new AtomicInteger();
+        final int failTimes = 6;
+        injectMockReplicatorProducerBuilder((producerCnf, orginalProducer) -> {
+            if (topicName.equals(producerCnf.getTopicName())) {
+                // There is a switch to determine create producer successfully or not.
+                if (createProducerCounter.incrementAndGet() > failTimes) {
+                    return orginalProducer;
+                }
+                log.info("Retry create replicator.producer count: {}", createProducerCounter);
+                // Release producer and fail callback.
+                orginalProducer.closeAsync();
+                throw new RuntimeException("mock error");
+            }
+            return orginalProducer;
+        });
+
+        // Create topic.
+        admin1.topics().createNonPartitionedTopic(topicName);
+        PersistentTopic persistentTopic =
+                (PersistentTopic) pulsar1.getBrokerService().getTopic(topicName, false).join().get();
+        PersistentReplicator replicator =
+                (PersistentReplicator) persistentTopic.getReplicators().values().iterator().next();
+        // Since we inject a producer creation error, the replicator can not start successfully.
+        assertFalse(replicator.isConnected());
+
+        // Stuck the closing of the cursor("pulsar.repl"), until the internal producer of the replicator started.
+        SpyCursor spyCursor =
+                spyCursor(persistentTopic, "pulsar.repl." + pulsar2.getConfig().getClusterName());
+        CursorCloseSignal cursorCloseSignal = makeCursorClosingDelay(spyCursor);
+
+        // Unload bundle: call "topic.close(false)".
+        // Stuck start new producer, until the state of replicator change to Stopped.
+        // The next once of "createProducerSuccessAfterFailTimes" to create producer will be successfully.
+        Awaitility.await().pollInterval(Duration.ofMillis(100)).atMost(Duration.ofSeconds(60)).untilAsserted(() -> {
+            assertTrue(createProducerCounter.get() >= failTimes);
+        });
+        CompletableFuture<Void> topicCloseFuture = persistentTopic.close(true);
+        Awaitility.await().atMost(Duration.ofSeconds(30)).untilAsserted(() -> {
+            String state = String.valueOf(replicator.getState());
+            assertTrue(state.equals("Stopped") || state.equals("Terminated"));
+        });
+
+        // Delay close cursor, until "replicator.producer" create successfully.
+        // The next once retry time of create "replicator.producer" will be 3.2s.
+        Thread.sleep(4 * 1000);
+        log.info("Replicator.state: {}", replicator.getState());
+        cursorCloseSignal.startClose();
+        cursorCloseSignal.startCallback();
```

Comment:
Why do we need to close the cursor with a delay? To make sure the replicator can read messages?

Reply:
> Why do we need to close the cursor with a delay? To make sure the Replicator can read messages?

Just make sure the `replicator_producer` will not be closed due to the cursor is closed.

- The next once-retry time of creating `replicator_producer` will be 3.2s.
- Once the `replicator_producer` is created, Replicator will do `readMoreEntries`
  - `cursor is closed`: Replicator will get a `readEntriesFailed` and close the `replicator_producer.`
  - `cursor is available`: Cursor will pend the reading requests because there are no entries in the backlog.

Reply:
Ok, but why the producer will not be closed in this case?
<img width="914" alt="image" src="https://github.com/apache/pulsar/assets/12592133/73ef2bbd-778e-4054-9b1e-7dad439ef282">

It should be finally closed?

Reply:
> It should be finally closed?

As expected, it will. But the current implementation of `cursor.asyncReadEntries(cb, ctx)` does not guarantee the `callback` will finally get a call 😂 

Cursor will pend the reading requests because there are no entries in the backlog and drop all pending reading requests when closing.